### ISSUE TITLE
Multilabel implementation

### ISSUE THREAD

**Issue by Jopepato** [dmlc/xgboost#3860]
Hi, thanks for this marvellous implementation of the algorithm.
I was considering using it for multilabel purposes but as I can see there is no implementation for it.
Considering the algorithm as treating the labels separately as a BinaryRelevance or even OVO. Should it be implemented inside the library or use it outside, using as a base classifier this xgboost

If you are considering implementing it inside, it would be an honour to make the PR! ^

**Comment by hcho3** [comment]
@Jopepato Sure, feel free to submit a pull request. We will gladly consider it.

**Comment by Jopepato** [comment]
@hcho3 It would be in the python implementation as it is in the language I usually work with. I was checking the code and thinking of doing this implementation in the _fit_ function directly in sklearn.py. Checking the shape of _y_ as the multilabel implementation will be an array of classifiers for each label of the _y_ set
Or maybe making another function that will call this fit and call it like fit_multilabel or something like this.
Which one will be the best option?

**Comment by hcho3** [comment]
@Jopepato Let's create a new function so that we don't break existing fit() function.

**Comment by thvasilo** [comment]
I'm wondering if it will be somehow possible to maintain the sklearn API and do multi-label as [sklearn does it](https://scikit-learn.org/stable/modules/multiclass.html).

They use an `X times |Y|` representation for the prediction matrix, and I assume for the data as well. If we want to maintain the users' expectation that XGBoost works with big data it will be hard to achieve for multi-label problems.

**Comment by Jopepato** [comment]
I was thinking of that as well. Training an X numbers of classifiers one for each label |Y|. But as you say the complexity and time for the model to train all the data would increase a lot...
Another approach would be as the LabelPowerset does... this is changing the multilabel output to a multiclass problem. Assigning each set of different labels to a class.
By doing this we lose some information as some set of labels would be unique (mostly in datasets with a high number of labels) but the computational time for the algorithm would be the same as for a multiclass problem.

**Comment by khotilov** [comment]
Did you consider using any existing packages that already implement various multilabel problem reduction strategies for a generic classifier? E.g., scikit-multilearn for Python or utiml for R.

Unless there would be some clear benefit in custom implementation of some of such strategies within xgboost, it would not make much sense to add and maintain such functionality here.

**Comment by Jopepato** [comment]
@hcho3 I guess that is a better option... using xgboost as the base classifier and letting other library work with the multi label problem...

**Comment by alexvpickering** [comment]
[Here](http://www.ke.tu-darmstadt.de/lehre/arbeiten/bachelor/2017/Bohlender_Simon.pdf) is a paper demonstrating an approach that was better than chained classifiers with xgboost.

**Comment by purviprajapati196** [comment]
kindly inform me is it applicable to multi-label domain or not.

**Comment by hcho3** [comment]
Closing for now.
